### ISSUE TITLE
Add some Powershell example (please!)

### ISSUE THREAD

**Issue by ilgrank** [dotnet/MQTTnet#1330]
Hi undestand it can be trivial for someone, but not for everyone.
I've searched long and not even a single hit returns on Google for using MQTTnet in Powershell
Could you please add some basic example?
Many thanks!

**Comment by chkr1011** [comment]
This library is no PowerShell Module. So working with this library in PowerShell works the same way as working with other .NET Code in PowerShell (https://stackoverflow.com/questions/3079346/how-to-reference-net-assemblies-using-powershell). 

Apart from that I cannot help much because I do not use PowerShell that often.

**Comment by ilgrank** [comment]
using the 4.0 Beta, I've come this far:

```
Add-Type -Path "C:\temp\MQTTnet.dll"
$Factory=[MQTTnet.MqttFactory]::new()
$MqttClient=$Factory.CreateMqttClient()
$Options=[MQTTnet.Client.MqttClientOptionsBuilder]::new()
$Options=[MQTTnet.Client.MqttClientOptionsBuilder]::new().WithClientId("Client1").WithTcpServer("test.mosquitto.org").WithCleanSession().Build()
$MqttClient.ConnectAsync($options, [System.Threading.CancellationToken]::None)

```
but then I try to subscribe to a topic, i.e.:
`$MqttClient.SubscribeAsync([MQTTnet.Client.MqttClientSubscribeOptionsBuilder]::new().WithTopic('Sample').Build())`
I cannot proceed since 
[MQTTnet.Client.MqttClientSubscribeOptionsBuilder]::new()
has only these methods:
```
  TypeName: MQTTnet.Client.MqttClientSubscribeOptions

Name                   MemberType Definition
----                   ---------- ----------
Equals                 Method     bool Equals(System.Object obj)
GetHashCode            Method     int GetHashCode()
GetType                Method     type GetType()
ToString               Method     string ToString()
SubscriptionIdentifier Property   uint32 SubscriptionIdentifier {get;set;}
TopicFilters           Property   System.Collections.Generic.List[MQTTnet.Packets.MqttTopicFilter] TopicFilters {get;set;}
UserProperties         Property   System.Collections.Generic.List[MQTTnet.Packets.MqttUserProperty] UserProperties {get;set;}
```

**Comment by diputssi** [comment]
I was able to get a bit further, so I'm adding additional steps here.  Looks like you can use the with topic filter method instead of WithTopic.

$topic = [MQTTnet.MqttTopicFilterBuilder]::new().WithTopic('testnode/#').build()

Then create a subscribe options object with the topic filter.

$SubscribeOptions = [MQTTnet.Client.MqttClientSubscribeOptionsBuilder]::new().WithTopicFilter( $topic ).Build()

Then I attempt to get at the awaiter object, but it finishes before I can get await on the results.

Sometimes you can see the awaiter not finish.

$awaiter = $MqttClient.SubscribeAsync($SubscribeOptions).getawaiter(); $Awaiter;
$Awaiter

Basically, I don't know how to do await tasks in PowerShell without having the SubscribeAsync trigger and finish before I set up the callback.  Maybe someone else can help us in setting up the callback first.

I currently get around all of this .Net nonsense by using a PowerShell wrapper around a python object based on paho.  But of course less than ideal.

**Comment by ilgrank** [comment]
To anyone ending here, seeking to use the only (as today) actively maintained .NET MQTT library.
Is kind of possible to use MQTTNet in powershell, but (afaik) there's no way to register the message events, hence you will be able to *publish* messages into topics, but not be able to *receive* them
That said, the following code works (tested in MQTTnet 4.1.x, Powershell 5.1, Win 10. Download compiled dlls from Nuget)
```
Add-Type -Path .\MQTTnet.dll
$MQTTClient=[MQTTnet.MqttFactory]::new().CreateMqttClient()
$ClientOptions=[MQTTnet.Client.MqttClientOptionsBuilder]::new().WithClientId("Client1").WithTcpServer("test.mosquitto.org").WithCleanSession().Build()
$MQTTClient.ConnectAsync($ClientOptions, [System.Threading.CancellationToken]::None)
$SubscribeOptions=[MQTTnet.Client.MqttClientSubscribeOptionsBuilder]::new().WithTopicFilter('#').Build()
$MQTTClient.SubscribeAsync($SubscribeOptions,[System.Threading.CancellationToken]::None)
$Message=[MQTTnet.MqttApplicationMessageBuilder]::new().WithTopic("test").WithPayload("hello world!").Build()
$MQTTClient.PublishAsync($Message)

```

**Comment by chkr1011** [comment]
We may create a proper PowerShell module. It may be way easier to use then importing importing this library manually.

**Comment by ilgrank** [comment]
> We may create a proper PowerShell module. It may be way easier to use then importing importing this library manually.

That would be *awesome*, as currently there's no viable alternative for Powershell users. I'm keeping my fingers crossed :)

**Comment by ilgrank** [comment]
> We may create a proper PowerShell module. It may be way easier to use then importing importing this library manually.

Sorry to necro this issue, has there been any update on this matter?

**Comment by hiroki-tyffon** [comment]
ROO CLINE wrote all this code.
It's a C# wrapper.
I'm going to test it now to see if I'm happy with it.

```Import-Module "$PSScriptRoot\..\..\env.psm1"

<#
.SYNOPSIS
A PowerShell MQTT client implementation using MQTTnet.dll.

.DESCRIPTION
This script demonstrates how to use MQTTnet.dll in PowerShell to create a fully functional MQTT client.
Features include:
- MQTT 5.0 protocol support
- QoS control (0, 1, 2)
- Retain flag support
- User Properties support
- PowerShell event-based message handling

.NOTES
Requirements:
- MQTTnet.dll (version 4.3.7.1207 or later)
- MQTT broker with MQTT 5.0 support

.EXAMPLE
# Set the MQTT broker host in environment variable
$env:MQTT_BROKER_HOST = "localhost"

# Run the script
.\Test-MqttNet.ps1

This will connect to the MQTT broker, subscribe to test topics, and demonstrate message publishing with QoS, Retain, and User Properties.
#>

# Path to MQTTnet.dll
$dllPath = "$rootDir\Lib\mqttnet.4.3.7.1207\lib\net48\MQTTnet.dll"

# Unblock the DLL if needed
Unblock-File -Path $dllPath

# Load the DLL
Add-Type -Path $dllPath

# Define the message handler class
$source = @"
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Protocol;
using MQTTnet.Packets;

/// <summary>
/// Wrapper class for MQTTnet client with enhanced PowerShell compatibility
/// </summary>
public class MqttClientWrapper : IDisposable
{
    private IMqttClient _client;
    private MqttClientOptions _options;
    
    public event EventHandler<MqttMessageEventArgs> MessageReceived;

    /// <summary>
    /// Event arguments for MQTT message events
    /// </summary>
    public class MqttMessageEventArgs : EventArgs
    {
        public string Topic { get; private set; }
        public string Payload { get; private set; }
        public MqttQualityOfServiceLevel QoS { get; private set; }
        public bool Retain { get; private set; }
        public IDictionary<string, string> UserProperties { get; private set; }

        public MqttMessageEventArgs(string topic, string payload, MqttQualityOfServiceLevel qos, bool retain, IDictionary<string, string> userProperties)
        {
            Topic = topic;
            Payload = payload;
            QoS = qos;
            Retain = retain;
            UserProperties = userProperties ?? new Dictionary<string, string>();
        }
    }

    /// <summary>
    /// Creates a new MQTT client wrapper
    /// </summary>
    /// <param name="host">MQTT broker hostname</param>
    /// <param name="port">MQTT broker port</param>
    public MqttClientWrapper(string host, int port)
    {
        var factory = new MqttFactory();
        _client = factory.CreateMqttClient();
        
        var optionsBuilder = new MqttClientOptionsBuilder()
            .WithTcpServer(host, port)
            .WithClientId("PowerShellClient")
            .WithProtocolVersion(MQTTnet.Formatter.MqttProtocolVersion.V500); // Use MQTT 5.0
        _options = optionsBuilder.Build();

        _client.ApplicationMessageReceivedAsync += HandleMessage;
    }

    private async Task HandleMessage(MqttApplicationMessageReceivedEventArgs e)
    {
        var topic = e.ApplicationMessage.Topic;
        var payload = System.Text.Encoding.UTF8.GetString(e.ApplicationMessage.PayloadSegment.Array);
        var qos = e.ApplicationMessage.QualityOfServiceLevel;
        var retain = e.ApplicationMessage.Retain;
        var userProperties = new Dictionary<string, string>();
        
        if (e.ApplicationMessage.UserProperties != null)
        {
            foreach (var prop in e.ApplicationMessage.UserProperties)
            {
                userProperties[prop.Name] = prop.Value;
            }
        }
        
        if (MessageReceived != null)
        {
            MessageReceived(this, new MqttMessageEventArgs(topic, payload, qos, retain, userProperties));
        }
        
        await Task.CompletedTask;
    }

    /// <summary>
    /// Connects to the MQTT broker
    /// </summary>
    public async Task ConnectAsync()
    {
        await _client.ConnectAsync(_options);
    }

    /// <summary>
    /// Subscribes to an MQTT topic
    /// </summary>
    /// <param name="topic">Topic to subscribe to (supports wildcards)</param>
    public async Task SubscribeAsync(string topic)
    {
        var optionsBuilder = new MqttClientSubscribeOptionsBuilder()
            .WithTopicFilter(topic);
        var options = optionsBuilder.Build();
        await _client.SubscribeAsync(options);
    }

    /// <summary>
    /// Publishes a message to an MQTT topic
    /// </summary>
    /// <param name="topic">Topic to publish to</param>
    /// <param name="payload">Message payload</param>
    /// <param name="qos">Quality of Service level (0, 1, or 2)</param>
    /// <param name="retain">Whether to retain the message</param>
    /// <param name="userProperties">Optional user properties (key-value pairs)</param>
    public async Task PublishAsync(string topic, string payload, MqttQualityOfServiceLevel qos = MqttQualityOfServiceLevel.AtMostOnce, bool retain = false, IDictionary<string, string> userProperties = null)
    {
        var builder = new MqttApplicationMessageBuilder()
            .WithTopic(topic)
            .WithPayload(payload)
            .WithQualityOfServiceLevel(qos)
            .WithRetainFlag(retain);

        if (userProperties != null)
        {
            foreach (var prop in userProperties)
            {
                builder.WithUserProperty(prop.Key, prop.Value);
            }
        }

        var message = builder.Build();
        await _client.PublishAsync(message);
    }

    /// <summary>
    /// Disconnects from the MQTT broker
    /// </summary>
    public async Task DisconnectAsync()
    {
        if (_client.IsConnected)
        {
            await _client.DisconnectAsync();
        }
    }

    public void Dispose()
    {
        if (_client != null)
        {
            _client.Dispose();
            _client = null;
        }
    }

    /// <summary>
    /// Gets whether the client is currently connected
    /// </summary>
    public bool IsConnected
    {
        get { return _client != null && _client.IsConnected; }
    }
}
"@

Add-Type -TypeDefinition $source -ReferencedAssemblies $dllPath

# Get MQTT broker hostname from environment variable
$mqttHost = $env:MQTT_BROKER_HOST
if (-not $mqttHost) {
    Write-Host "Error: MQTT_BROKER_HOST environment variable is not set"
    exit 1
}

try {
    # Create MQTT client
    $client = New-Object MqttClientWrapper($mqttHost, 1883)

    # Register PowerShell event handler
    Register-ObjectEvent -InputObject $client -EventName MessageReceived -Action {
        $topic = $EventArgs.Topic
        $payload = $EventArgs.Payload
        $qos = $EventArgs.QoS
        $retain = $EventArgs.Retain
        $userProps = $EventArgs.UserProperties
        
        Write-Host "Received message: Topic=$topic, Payload=$payload, QoS=$qos, Retain=$retain"
        if ($userProps.Count -gt 0) {
            Write-Host "User Properties:"
            foreach ($prop in $userProps.GetEnumerator()) {
                Write-Host "  $($prop.Key): $($prop.Value)"
            }
        }
    } | Out-Null

    # Connect to broker
    Write-Host "Connecting to MQTT broker ($mqttHost)..."
    $client.ConnectAsync().GetAwaiter().GetResult()
    Write-Host "Connected successfully"

    # Subscribe to topic
    $topic = "test/powershell/#"
    Write-Host "Subscribing to topic '$topic'..."
    $client.SubscribeAsync($topic).GetAwaiter().GetResult()
    Write-Host "Subscribed successfully"

    # Publish a message with QoS 1, Retain flag, and User Properties
    $pubTopic = "test/powershell/message"
    $payload = "Hello from PowerShell!"
    $userProps = New-Object 'System.Collections.Generic.Dictionary[string,string]'
    $userProps.Add("sender", "powershell-client")
    $userProps.Add("timestamp", [DateTime]::UtcNow.ToString("o"))
    
    Write-Host "Publishing message... (QoS=1, Retain=true, with User Properties)"
    $client.PublishAsync($pubTopic, $payload, [MQTTnet.Protocol.MqttQualityOfServiceLevel]::AtLeastOnce, $true, $userProps).GetAwaiter().GetResult()
    Write-Host "Message published: $payload"

    # Wait for messages
    Write-Host "Waiting for messages... (Press Ctrl+C to exit)"
    while ($true) {
        Start-Sleep -Seconds 1
    }

} catch {
    Write-Host "Error occurred: $_"
} finally {
    # Unregister event handler
    Get-EventSubscriber | Where-Object { $_.SourceObject -eq $client } | Unregister-Event
    
    # Disconnect
    if ($client -and $client.IsConnected) {
        Write-Host "Disconnecting..."
        $client.DisconnectAsync().GetAwaiter().GetResult()
    }
    if ($client) {
        $client.Dispose()
    }
}
```
